java-util/src/main/java/io/druid/java/util/common/io/NativeIO.java

```diff
@@ -0,0 +1,168 @@
+  private NativeIO() {}
+
+  private static Field getFieldByReflection(Class cls, String fieldName)
+  {
+    Field field = null;
+
+    try {
+      field = cls.getDeclaredField(fieldName);
+      field.setAccessible(true);
+    }
+    catch (Exception e) {
+      // We don't really expect this so throw an assertion to
+      // catch this during development
+      assert false;
+      LOG.warn("Unable to read {} field from {}", fieldName, cls.getName());
+    }
+
+    return field;
+  }
+  /**
+   * Get system file descriptor (int) from FileDescriptor object.
+   * @param descriptor - FileDescriptor object to get fd from
+   * @return file descriptor, -1 or error
+   */
+  public static int getfd(FileDescriptor descriptor)
+  {
+    Field field = getFieldByReflection(descriptor.getClass(), "fd");
+    try {
+      return field.getInt(descriptor);
+    }
+    catch (Exception e) {
+      LOG.warn("Unable to read fd field from java.io.FileDescriptor");
+    }
+
+    return -1;
+  }
+
+  /**
+   * Remove pages from the file system page cache when they wont
+   * be accessed again
+   *
+   * @param fd     The file descriptor of the source file.
+   * @param offset The offset within the file.
+   * @param len    The length to be flushed.
+   */
+
+  public static void trySkipCache(int fd, long offset, long len)
+  {
+    if (!initialized || !fadvisePossible || fd < 0) {
+      return;
+    }
+    try {
+      posix_fadvise(fd, offset, len, POSIX_FADV_DONTNEED);
+    }
+    catch (UnsupportedOperationException uoe) {
+      LOG.warn("posix_fadvise is not supported : ", uoe);
+      fadvisePossible = false;
+    }
+    catch (UnsatisfiedLinkError ule) {
+      // if JNA is unavailable just skipping Direct I/O
+      // instance of this class will act like normal RandomAccessFile
+      LOG.warn("Unsatisfied Link error: posix_fadvise failed on file descriptor {}, offset {} : ",
+          new Object[] {fd, offset, ule});
+      fadvisePossible = false;
+    }
+    catch (Exception e) {
+      // This is best effort anyway so lets just log that there was an
+      // exception and forget
+      LOG.warn("Unknown exception: posix_fadvise failed on file descriptor {}, offset {} : ",
+          new Object[] {fd, offset, e});
+    }
+  }
+
+
+  /**
+   * Copy from an input stream to a file minimizing cache impact. This happens chunk by chunk
+   * so only at most chunk size will be present in the OS page cache. Linux only.
+   *
+   * @param src Source InputStream where to copy from
+   * @param dest Destination file to copy to
+   * @throws IOException
+   */
+  public static void chunkedCopy(InputStream src, File dest) throws IOException
+  {
+    final RandomAccessFile raf = new RandomAccessFile(dest, "rwd");
+    final int fd = getfd(raf.getFD());
+
+    final byte[] buf = new byte[1024 * 1024 * 8]; // 8Mb buffer
+    int numBytes;
+    long offset = 0;
+
+    LOG.info("Started chunked copy to " + dest);
+
+    while ((numBytes = src.read(buf)) > 0) {
```

Comment:
It has been a while since I dug deeply into input stream contracts. Had to check this one,

`java.io.InputStream#read(byte[])` says either `-1` is returned or a positive byte count. It blocks otherwise, so this should be reasonable logic. so seems good

Reply:
What I need to look into is that the reading from the stream can happen in very small chunks (couple of bytes it seems) I don't think that should impact what the trySkipCache is doing, but given the fact integration tests are not passing it might be an issues. So maybe I will wait until the buffer is full or the stream has finished before writing out and see if that solves the passing of the tests.

Reply:
it can be nasty if the target is set to `rwd` because the sync on many tiny chunks will be bad.